core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala

```diff
@@ -1609,6 +1609,16 @@ class TestFileSystem extends org.apache.hadoop.fs.LocalFileSystem {
   }
 
   override def open(path: Path): FSDataInputStream = super.open(local(path))
+
+  // No-op methods
+
+  override def copyFromLocalFile(
```

Comment:
+1 on the idea. Why we need override methods `copyFromLocalFile` and `mkdirs` here?

Reply:
This was needed for the unit test to work. I was looking for a way to mock the [KubernetesUtils.uploadFileUri()](https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala#L310-L335) method which is used to actually upload the local files to a remote path specified in `spark.kubernetes.file.upload.path`. `KubernetesUtils` is an `object` which are notoriously hard to mock and there was a bit too much refactoring required to make this method mockable.

So I decided to go with another approach: use a fake Hadoop FS so that the operations performed in the `uploadFileUri()` method become no-ops. I found this `TestFileSystem` class which was used in other tests for the same purpose so I simply added the methods which were called inside `uploadFileUri()`: `mkdirs` [here](https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala#L320) and `copyFromLocalFile` [here](https://github.com/apache/spark/blob/a27b459be3ca2ad2d50b9d793b939071ca2270e2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala#L347). I then made sure that this FS implementation was used by the test by specifying it here: https://github.com/antonipp/spark/blob/80416f6a6ff42051cd38dfe32df2d0c7cd0741f4/resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStepSuite.scala#L358

Reply:
nit: Each level of indentation is 2 spaces in apache spark.  https://docs.scala-lang.org/style/indentation.html  
BTW, if you want to test with `KubernetesUtils`, you could add integration test with MINIO in `DepsTestsSuite`.

Reply:
Fixed the indentation in a new commit.

I also spent some time trying to make integration tests work on my machine but encountered a blocking issue with my local setup, I opened a ticket here: https://issues.apache.org/jira/browse/SPARK-40954 
Could we merge this PR as it is? I can spend more time on fixing my integration test setup + writing a new test some time later but I think the unit test I added already gives us pretty good coverage.

Reply:
Thanks. I have left comments on [SPARK-40954](https://issues.apache.org/jira/browse/SPARK-40954). I'm not sure if this issue is only on the M1, I haven't reproduced the issue on my Intel Mac. Maybe try using `hyperkit` first to see if the issue recurs. This PR needs others to review, let me @dongjoon-hyun @holdenk

Reply:
Well, it took me a a bit of time to get to it but I finally have a working local setup and I wrote an integration test for this PR. I added it in https://github.com/apache/spark/pull/38376/commits/cbda5403b00f4e66ad4651ceda7786de1f5e1e1a

It worked for me locally with:
```bash
mvn integration-test -am -pl :spark-kubernetes-integration-tests_2.12 \
                        -Pkubernetes -Pkubernetes-integration-tests -Phadoop-3 \
                        -Dspark.kubernetes.test.sparkTgz=/path/to/the/tgz/built/from/this/branch \
                        -Dspark.kubernetes.test.deployMode=minikube \
                        -Dspark.kubernetes.test.imageRepo=docker.io/kubespark \
                        -Dspark.kubernetes.test.namespace=spark \
                        -Dspark.kubernetes.test.serviceAccountName=spark
```

@dongjoon-hyun @holdenk, would appreciate your reviews as well 🙏